sphinx_gallery/recommender.py

```diff
@@ -0,0 +1,273 @@
+            max_df = int(np.floor(max_df * counts_matrix.shape[0]))
+        elif not isinstance(max_df, numbers.Integral) and max_df > 1:
+            raise ValueError("max_df must be float in range [0.0, 1.0] or int")
+        doc_appearances = np.sum(counts_matrix, axis=0)
+        mask = (doc_appearances >= min_df) & (doc_appearances <= max_df)
+        self.similarity_matrix_ = self.cosine_similarity(
+            self.compute_tf_idf(counts_matrix[:, mask])
+        )
+        self.file_names_ = file_names
+        return self
+
+    def predict(self, file_name):
+        """Compute the most `n_examples` similar documents to the query.
+
+        Parameters
+        ----------
+        file_name : str
+            Name of the file corresponding to the query index `item_id`.
+
+        Returns
+        -------
+        recommendations : list of str
+            Name of the files most similar to the query.
+        """
+        item_id = self.file_names_.index(file_name)
+        similar_items = list(enumerate(self.similarity_matrix_[item_id]))
+        sorted_items = sorted(similar_items, key=lambda x: x[1], reverse=True)
+
+        # Get the top k items similar to item_id. Note that `sorted_items[0]`
+        # is always the query document itself, hence it is discarded from the
+        # returned list of recommendations.
+        top_k_items = [idx for idx, _ in sorted_items[1 : self.n_examples + 1]]
+        recommendations = [self.file_names_[idx] for idx in top_k_items]
+        return recommendations
+
+
+def _write_recommendations(recommender, fname, gallery_conf):
+    """Generate `.recommendations` RST file for a given example.
+
+    Parameters
+    ----------
+    recommender : ExampleRecommender
+        Instance of a fitted ExampleRecommender.
+
+    fname : str
+        Path to the example file.
+
+    gallery_conf : dict
+        Configuration dictionary for the sphinx-gallery extension.
+    """
+    path_fname = Path(fname)
+    recommend_fname = f"{path_fname.parent / path_fname.stem}.recommendations.new"
+    recommended_examples = recommender.predict(fname)
+
+    with open(recommend_fname, "w", encoding="utf-8") as ex_file:
+        ex_file.write("\n\n.. rubric:: Related examples\n")
+        ex_file.write(THUMBNAIL_PARENT_DIV)
+        for example_fname in recommended_examples:
+            example_path = Path(example_fname)
+            _, script_blocks = split_code_and_text_blocks(
+                example_fname, return_node=False
+            )
+            intro, title = extract_intro_and_title(fname, script_blocks[0][1])
+            ex_file.write(
+                _thumbnail_div(
+                    example_path.parent,
+                    gallery_conf["src_dir"],
+                    example_path.name,
+                    intro,
+                    title,
+                    is_backref=True,
+                )
+            )
+        ex_file.write(THUMBNAIL_PARENT_DIV_CLOSE)
```

Comment:
We could also insert of user customizable rubric footer here. E.g. to link to API doc of the recommender system on the sphinx-gallery documentation website.